### PR TITLE
feat: add provisioning timings to understand slow build times

### PR DESCRIPTION
**Pull request by dannykopping**
This PR productionizes a spike I conducted during https://github.com/coder/team-coconut/issues/22.

Terraform exposes discrete timing information for each resource when planning and applying, and we're now using this to build up a view of which resources or stages take the most time. I'm hooking into our existing parsing of the Terraform log lines and expanding it a bit; there were some necessary refactors but I've tried to keep it as low-touch as possible.

This PR does _not_ contain any HTTP APIs or UIs. Those will have to be added in subsequent PRs; this one is already fairly large.

There are tests, and they cover the majority of the new code added, but they're incomplete. I had to do some fairly nasty business in `provisioner/terraform/timings_test.go` since we don't have a seam in front of calls to `os/exec` which makes it difficult to mock. There's some precedent for this though in `provisioner/terraform/provision_test.go` from which I took heavy inspiration.

**We also don't have _any_ active e2e tests which use terraform, which I think is something worth addressing.**

I also included `txtar` after watching Russ Cox's [excellent talk on Go testing](https://research.swtch.com/testing) recently, and it made the tests quite readable - at least I think so. If you disagree, let me know.

---

Using this information, we'll be able to add cool graphs like this to workspace detail pages and probably also expose some aggregated data in Prometheus metrics.

image

### CODE REVIEW

File: coderd/database/migrations/000245_provisioner_job_timings.up.sql [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+           MIN(CASE WHEN pjt.stage = 'init'::provisioner_job_timing_stage THEN pjt.started_at END))), 0)  AS init_secs,
+       GREATEST(EXTRACT(EPOCH FROM (
+           MAX(CASE WHEN pjt.stage = 'plan'::provisioner_job_timing_stage THEN pjt.ended_at END) -
+           MIN(CASE WHEN pjt.stage = 'plan'::provisioner_job_timing_stage THEN pjt.started_at END))), 0)  AS plan_secs,
+       GREATEST(EXTRACT(EPOCH FROM (
+           MAX(CASE WHEN pjt.stage = 'graph'::provisioner_job_timing_stage THEN pjt.ended_at END) -
+           MIN(CASE WHEN pjt.stage = 'graph'::provisioner_job_timing_stage THEN pjt.started_at END))), 0) AS graph_secs,
+       GREATEST(EXTRACT(EPOCH FROM (
+           MAX(CASE WHEN pjt.stage = 'apply'::provisioner_job_timing_stage THEN pjt.ended_at END) -
+           MIN(CASE WHEN pjt.stage = 'apply'::provisioner_job_timing_stage THEN pjt.started_at END))), 0) AS apply_secs
```
Comment (dannykopping) [inline coderd/database/migrations/000245_provisioner_job_timings.up.sql:41]: Use the max(ended_at)-min(started_at) of each stage to determine stage timings.

File: coderd/workspacebuilds.go [hunk 2]
```diff
@@ -635,6 +635,7 @@ func (api *API) workspaceBuildState(rw http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// TODO: why?
```
Comment (dannykopping) [inline coderd/workspacebuilds.go:638]: Drive-by: I don't understand why this endpoint has to authorize a template?
Looks like copy-pasta.

File: provisioner/terraform/executor.go [hunk 3]
```diff
@@ -539,46 +556,42 @@ func readAndLog(sink logSink, r io.Reader, done chan<- any, level proto.LogLevel
 // provisionLogWriter creates a WriteCloser that will log each JSON formatted terraform log.  The WriteCloser must be
 // closed by the caller to end logging, after which the returned channel will be closed to indicate that logging of the
 // written data has finished.  Failure to close the WriteCloser will leak a goroutine.
-func provisionLogWriter(sink logSink) (io.WriteCloser, <-chan any) {
+func (e *executor) provisionLogWriter(sink logSink) (io.WriteCloser, <-chan any) {
```
Comment (dannykopping) [inline provisioner/terraform/executor.go:559]: Moved these under the `executor` receiver to get access to the logger.

File: provisioner/terraform/executor.go [hunk 4]
```diff
@@ -609,12 +676,25 @@ func convertTerraformLogLevel(logLevel string, sink logSink) proto.LogLevel {
 }
 
 type terraformProvisionLog struct {
-	Level   string `json:"@level"`
-	Message string `json:"@message"`
+	Level     string                    `json:"@level"`
```
Comment (dannykopping) [inline provisioner/terraform/executor.go:679]: I expanded this type to unmarshal additional fields from the existing JSON logs.

File: site/e2e/google/protobuf/timestampGenerated.ts [hunk 5]
```diff
@@ -0,0 +1,123 @@
+/* eslint-disable */
```
Comment (dannykopping) [inline site/e2e/google/protobuf/timestampGenerated.ts:1]: I don't understand:
1. why this file needs to be generated given that we use `google.protobuf.Timestamp` elsewhere (I guess the provisioner is used in e2e somehow?)
2. WTF the linter issues are.

Please help

File: provisioner/terraform/testutil.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
+package terraform
```
Comment (mafredri) [inline provisioner/terraform/testutil.go:1]: Should we name this `util_test.go` so it can't be imported from regular code?

File: provisioner/terraform/timings.go [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+// ingest accepts a timing span at a certain timestamp and assigns it a state according to the kind of timing event.
+// We memoize start & completion events, and then calculate their total duration in aggregate.
+// We ignore progress events because we only care about the full duration of the action (delta between *_start and *_complete events).
+func (t *timingAggregator) ingest(ts time.Time, s *timingSpan) {
+	if s == nil {
+		return
+	}
+
+	s.stage = t.stage
```
Comment (mafredri) [inline provisioner/terraform/timings.go:83]: No need for mutex lock while accessing `t` here? Might be good to document on the struct what the mutex protects, e.g.

```go
type timingAggregator struct {
	stage       database.ProvisionerJobTimingStage

	mu sync.Mutex // Protects following.
	stateLookup map[uint64]*timingSpan
}
```

File: provisioner/terraform/timings.go [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+}
+
+// aggregate performs a pass through all memoized events to build up a slice of *proto.Timing instances which represent
+// the total time taken to perform a certain action.
+// The resulting slice of *proto.Timing is NOT sorted.
+func (t *timingAggregator) aggregate() []*proto.Timing {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	out := make([]*proto.Timing, 0, len(t.stateLookup))
```
Comment (mafredri) [inline provisioner/terraform/timings.go:115]: I feel like this could be simplified, depending on how much we end up skipping the minor allocation saving is probably a wash?
```suggestion
	out := []*proto.Timing{}
```

File: coderd/workspacebuilds.go (hunk 2 above)
Comment (johnstcn) [reply coderd/workspacebuilds.go:638]: It's *not*. If you can get/push the workspace build state, you can potentially do nasty things. See: https://github.com/coder/security/issues/2

File: provisioner/terraform/timings_internal_test.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	//go:embed testdata/timings-aggregation/init.txtar
+	inputInit []byte
+	//go:embed testdata/timings-aggregation/error.txtar
+	inputError []byte
+	//go:embed testdata/timings-aggregation/complete.txtar
+	inputComplete []byte
+	//go:embed testdata/timings-aggregation/incomplete.txtar
+	inputIncomplete []byte
+	//go:embed testdata/timings-aggregation/faster-than-light.txtar
+	inputFasterThanLight []byte
```
Comment (mafredri) [inline provisioner/terraform/timings_internal_test.go:27]: How about using an `filepath.Glob` instead? Then new tests can be added by just dropping new files in the folder.

**Review by mafredri — COMMENTED** [review]
Nice work, some minor nits and comments but looks good in general 👍🏻

File: coderd/database/queries/provisionerjobs.sql [hunk 10]
```diff
@@ -144,3 +144,15 @@ WHERE
 	updated_at < $1
 	AND started_at IS NOT NULL
 	AND completed_at IS NULL;
+
+-- name: InsertProvisionerJobTimings :many
+INSERT INTO provisioner_job_timings (job_id, started_at, ended_at, stage, source, action, resource)
```
Comment (johnstcn) [inline coderd/database/queries/provisionerjobs.sql:149]: how many rows do we expect to have, on average, per build?

File: provisionerd/proto/provisionerd.pb.go [hunk 11]
```diff
@@ -1,7 +1,7 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.30.0
-// 	protoc        v4.23.3
+// 	protoc-gen-go v1.34.2
+// 	protoc        v4.25.3
```
Comment (johnstcn) [inline provisionerd/proto/provisionerd.pb.go:4]: was this intentional?

File: provisionerd/proto/provisionerd_drpc.pb.go [hunk 12]
```diff
@@ -1,5 +1,5 @@
 // Code generated by protoc-gen-go-drpc. DO NOT EDIT.
-// protoc-gen-go-drpc version: v0.0.33
+// protoc-gen-go-drpc version: (devel)
```
Comment (johnstcn) [inline provisionerd/proto/provisionerd_drpc.pb.go:2]: :eyes:

File: provisioner/terraform/timings.go [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+		return "state refresh"
+	default:
+		return "?"
+	}
+}
+
+// hashState computes a hash based on a timingSpan's unique properties and state.
+// The combination of resource and provider names MUST be unique across entries.
+func (e *timingSpan) hashByState(state proto.TimingState) uint64 {
+	id := fmt.Sprintf("%s:%s:%s:%s", e.kind.Category(), state.String(), e.resource, e.provider)
```
Comment (johnstcn) [inline provisioner/terraform/timings.go:201]: nit: strings.Builder?

File: site/e2e/google/protobuf/timestampGenerated.ts (hunk 5 above)
Comment (johnstcn) [reply site/e2e/google/protobuf/timestampGenerated.ts:1]: 1) `site/e2e` does use this.
2) You probably need to add it to `site/.eslintignore` and `site/.prettierignore` similar to what's already done for `provisionerGenerated.ts`

File: coderd/database/dbmem/dbmem.go [hunk 14]
```diff
@@ -6633,6 +6633,15 @@ func (q *FakeQuerier) InsertProvisionerJobLogs(_ context.Context, arg database.I
 	return logs, nil
 }
 
+func (*FakeQuerier) InsertProvisionerJobTimings(_ context.Context, arg database.InsertProvisionerJobTimingsParams) ([]database.ProvisionerJobTiming, error) {
+	err := validateDatabaseType(arg)
+	if err != nil {
+		return nil, err
+	}
+
+	return nil, nil
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/dbmem/dbmem.go:6642]: Didn't really seem necessary since I can't currently test the full roundtrip including calls to the db.
Happy to add it though if you want it for completeness' sake.

File: coderd/database/migrations/000245_provisioner_job_timings.up.sql [hunk 15]
```diff
@@ -0,0 +1,45 @@
+CREATE TYPE provisioner_job_timing_stage AS ENUM (
+    'init',
+    'plan',
+    'graph',
+    'apply'
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/migrations/000245_provisioner_job_timings.up.sql:5]: Probably not, no. I wanted a bit of type safety around the stages (as you'll see in the tests), but I could factor them out. In reality though we won't know how the abstractions bleed until we add another provisioner type, so I'd probably want to keep this for now.

File: coderd/database/migrations/000245_provisioner_job_timings.up.sql [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    );
+
+CREATE TABLE provisioner_job_timings
+(
+    job_id     uuid                         NOT NULL REFERENCES provisioner_jobs (id) ON DELETE CASCADE,
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/migrations/000245_provisioner_job_timings.up.sql:10]: Thought so too, but it seems to be the convention:

```bash
$ grep 'REFERENCES provisioner_jobs(id)' coderd/database/dump.sql 
    ADD CONSTRAINT parameter_schemas_job_id_fkey FOREIGN KEY (job_id) REFERENCES provisioner_jobs(id) ON DELETE CASCADE;
    ADD CONSTRAINT provisioner_job_logs_job_id_fkey FOREIGN KEY (job_id) REFERENCES provisioner_jobs(id) ON DELETE CASCADE;
    ADD CONSTRAINT provisioner_job_timings_job_id_fkey FOREIGN KEY (job_id) REFERENCES provisioner_jobs(id) ON DELETE CASCADE;
    ADD CONSTRAINT workspace_builds_job_id_fkey FOREIGN KEY (job_id) REFERENCES provisioner_jobs(id) ON DELETE CASCADE;
    ADD CONSTRAINT workspace_resources_job_id_fkey FOREIGN KEY (job_id) REFERENCES provisioner_jobs(id) ON DELETE CASCADE;
```

File: coderd/database/queries/provisionerjobs.sql (hunk 10 above)
Comment (dannykopping) [reply coderd/database/queries/provisionerjobs.sql:149]: How long is a piece of string? 😛 
Depends on the number of resources. I'll pull some numbers from the existing logs next week to be sure, but I'd imagine it'll be in the 10-100 range.

File: coderd/database/queries/provisionerjobs.sql [hunk 17]
```diff
@@ -144,3 +144,15 @@ WHERE
...
+SELECT
+    @job_id::uuid AS provisioner_job_id,
+    unnest(@started_at::timestamptz[]) AS started_at,
+    unnest(@ended_at::timestamptz[]) AS ended_at,
+    unnest(@stage::provisioner_job_timing_stage[]) AS context,
+    unnest(@source::text[]) AS context,
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/queries/provisionerjobs.sql:155]: Double dumb.
Thanks, bad copy-pasta.

[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/queries/provisionerjobs.sql:155]: Luckily the code-gen's got my back:

```go
InsertProvisionerJobTimings(ctx context.Context, arg database.InsertProvisionerJobTimingsParams) ([]database.ProvisionerJobTiming, error) {
...

type ProvisionerJobTiming struct {
	JobID     uuid.UUID                 `db:"job_id" json:"job_id"`
	StartedAt time.Time                 `db:"started_at" json:"started_at"`
	EndedAt   time.Time                 `db:"ended_at" json:"ended_at"`
	Stage     ProvisionerJobTimingStage `db:"stage" json:"stage"`
	Source    string                    `db:"source" json:"source"`
	Action    string                    `db:"action" json:"action"`
	Resource  string                    `db:"resource" json:"resource"`
}
```

File: coderd/database/sqlc.yaml [hunk 18]
```diff
@@ -73,6 +73,9 @@ sql:
           - column: "notification_messages.payload"
             go_type:
               type: "[]byte"
+          - column: "provisioner_job_stats.*_secs"
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/sqlc.yaml:76]: Casts didn't work so I landed here; I took a change and globbing worked! 🎉

File: coderd/provisionerdserver/provisionerdserver.go [hunk 19]
```diff
@@ -1441,6 +1441,39 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
...
+			JobID: jobID,
+		}
+		for _, t := range completed.GetWorkspaceBuild().GetTimings() {
+			var start, end time.Time
+			if t.Start != nil {
+				start = t.Start.AsTime()
+			}
+			if t.End != nil {
+				end = t.End.AsTime()
+			}
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/provisionerdserver/provisionerdserver.go:1456]: I don't believe this is possible right now because in `aggregate()` I prevent returning entries like that, but you're right though I shouldn't rely on _spooky action at a distance._

File: coderd/provisionerdserver/provisionerdserver.go [hunk 20]
```diff
@@ -1441,6 +1441,39 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
...
+
+			var stg database.ProvisionerJobTimingStage
+			if err := stg.Scan(t.Stage); err != nil {
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/provisionerdserver/provisionerdserver.go:1459]: Precautionary.

File: coderd/provisionerdserver/provisionerdserver.go [hunk 21]
```diff
@@ -1441,6 +1441,39 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
...
+			params.Stage = append(params.Stage, stg)
+			params.Source = append(params.Source, t.Source)
+			params.Resource = append(params.Resource, t.Resource)
+			params.Action = append(params.Action, t.Action)
+			params.StartedAt = append(params.StartedAt, start)
+			params.EndedAt = append(params.EndedAt, end)
+		}
+		_, err = s.Database.InsertProvisionerJobTimings(ctx, params)
+		if err != nil {
+			// Don't fail the transaction for non-critical data.
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/provisionerdserver/provisionerdserver.go:1473]: See the comment at the top of this block 😝 
But no.

File: coderd/workspacebuilds.go (hunk 2 above)
Comment (dannykopping) [reply coderd/workspacebuilds.go:638]: Oooh, TIL!

File: provisioner/terraform/executor.go [hunk 22]
```diff
@@ -590,6 +603,60 @@ func provisionReadAndLog(sink logSink, r io.Reader, done chan<- any) {
...
+
+	typ := timingKind(log.Type)
+	if !typ.Valid() {
+		return time.Time{}, nil, xerrors.Errorf("unexpected timing kind: %q", log.Type)
+	}
+
+	ts, err := time.Parse("2006-01-02T15:04:05.000000Z07:00", log.Timestamp)
+	if err != nil {
+		// TODO: log
+		ts = time.Now()
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply provisioner/terraform/executor.go:649]: Deliberately didn't make this UTC because it gets run through `dbtime.Time` later on, but good thinking.

File: provisioner/terraform/testutil.go (hunk 6 above)
Comment (dannykopping) [reply provisioner/terraform/testutil.go:1]: Hhmm I tried that but ran into an issue; I think I wanted to use `util_test` but then Go complained that there were multiple package names. Looking at this again though I should definitely change this to `terraform_test` and be done with it.

File: provisioner/terraform/timings.go (hunk 7 above)
Comment (dannykopping) [reply provisioner/terraform/timings.go:83]: Good call. It's only to protect the map right now.

File: provisioner/terraform/timings.go (hunk 8 above)
Comment (dannykopping) [reply provisioner/terraform/timings.go:115]: Better to have ~loved and lost than never to have loved at all~ allocated a small bit of memory upfront then potentially incur multiple allocs later. IIRC slices start with a cap of 4 then double each time it's filled; if we have 100 items to manage that's ~5 allocs instead of 1, and the process of allocing is what's really slow AFAIK.

File: provisioner/terraform/timings_internal_test.go (hunk 9 above)
Comment (dannykopping) [reply provisioner/terraform/timings_internal_test.go:27]: I wanted to avoid too much magic. We also might want to add more members to the table test struct later which would then make it awkward.

File: site/e2e/google/protobuf/timestampGenerated.ts (hunk 5 above)
Comment (dannykopping) [reply site/e2e/google/protobuf/timestampGenerated.ts:1]: 1. Thanks!
2. Oh, I didn't add anything; this is all generated. I'l look into what's generating this.

Comment (johnstcn) [reply site/e2e/google/protobuf/timestampGenerated.ts:1]: It's `gen:provisioner` in `site/package.json`. It looks like `protoc-gen-ts_proto` automatically generates it with that blanket `eslint-ignore` directive. I wouldn't worry about that and simply exclude it from being linted as it is a generated file.

File: coderd/database/queries/provisionerjobs.sql (hunk 10 above)
Comment (johnstcn) [reply coderd/database/queries/provisionerjobs.sql:149]: Gotcha. We may need to think about pruning old workspace builds at some point; we currently have 32,631 rows in our dev db starting from 2022-08-10.

File: provisioner/terraform/testutil.go (hunk 6 above)
Comment (mafredri) [reply provisioner/terraform/testutil.go:1]: Ah, I was simply referring to renaming the _file_ from `testutil.go` to `util_test.go` (`_test.go` will exclude it from non-test code). Up to you if you want to change the package from `terraform` to `terraform_test` (internal vs external testing).

File: provisioner/terraform/timings.go (hunk 8 above)
Comment (mafredri) [reply provisioner/terraform/timings.go:115]: I'll leave it up to you.

That said, I don't think pre-allocating the slice will change allocs significantly considering we're already using a slice of pointers. I wrote a quick benchmark, and it may be a poor comparison, but this is the result:

```
❯ go test -bench .
goos: darwin
goarch: arm64
BenchmarkAlloc-10           	  750336	      1579 ns/op	    2976 B/op	     109 allocs/op
BenchmarkAllocUpfront-10    	  972052	      1210 ns/op	    1704 B/op	     102 allocs/op
PASS
```

https://go.dev/play/p/sQOsCPL4fW9

File: provisioner/terraform/timings.go (hunk 13 above)
Comment (dannykopping) [reply provisioner/terraform/timings.go:203]: Agreed but it's less concise and this is clear enough IMHO.

File: coderd/database/migrations/000245_provisioner_job_timings.up.sql [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    started_at timestamp with time zone     not null,
+    ended_at   timestamp with time zone     not null,
+    stage      provisioner_job_timing_stage not null,
+    source     text                         not null,
+    action     text                         not null,
+    resource   text                         not null
+);
+
+CREATE VIEW provisioner_job_stats AS
+SELECT pj.id                                                                                              AS job_id,
```
[the remark this answers is not in the document]
Comment (dannykopping) [reply coderd/database/migrations/000245_provisioner_job_timings.up.sql:20]: I just used my editor's formatter. I'll keep it like this because in the future it'll just get reverted if I changed it.

File: provisionerd/proto/provisionerd.pb.go (hunk 11 above)
Comment (dannykopping) [reply provisionerd/proto/provisionerd.pb.go:4]: Nope! Good catch.

Comment (dannykopping) [reply provisionerd/proto/provisionerd.pb.go:4]: Actually, we don't lock these versions.
In my dogfood workspace running nix, i've got:

```bash
$ protoc-gen-go --version
protoc-gen-go v1.34.2
$ protoc --version
libprotoc 25.3

$ which protoc protoc-gen-go
/nix/var/nix/profiles/default/bin/protoc
/nix/var/nix/profiles/default/bin/protoc-gen-go
```

Comment (johnstcn) [reply provisionerd/proto/provisionerd.pb.go]: I think CI uses the 'regular' dogfood image as opposed to 'nix'. This probably isn't a huge issue tbh but just wanted to double-check.

Comment (dannykopping) [reply provisionerd/proto/provisionerd.pb.go]: I would beg to differ.

If our dev environments' and our CI environment's dependencies are not in lock-step, this could lead not only to CI frustrations like the one above but to something more pernicious. We're either getting false-positives or false-negatives: any alignment is temporary/incidental, and we have no source of truth.

File: coderd/provisionerdserver/provisionerdserver.go [hunk 24]
```diff
@@ -1441,6 +1441,36 @@ func (s *server) CompleteJob(ctx context.Context, completed *proto.CompletedJob)
...
 		}
 
+		// Insert timings outside transaction since it is metadata.
+		// nolint:exhaustruct // The other fields are set further down.
+		params := database.InsertProvisionerJobTimingsParams{
+			JobID: jobID,
+		}
+		for _, t := range completed.GetWorkspaceBuild().GetTimings() {
+			if t.Start == nil || t.End == nil {
+				s.Logger.Warn(ctx, "timings entry has nil start or end time", slog.F("entry", t.String()))
```
Comment (johnstcn) [inline coderd/provisionerdserver/provisionerdserver.go:1451]: nit, non-blocking: Instead of logging the whole entry, could we instead log individual fields? I don't see anything sus in the fields right now, but it could lead us to log something later that we'd rather not.

File: provisioner/terraform/internal/timings_test_utils.go [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+
+		if a.Start.AsTime().Equal(b.Start.AsTime()) {
+			// Special case: when start times are equal, we need to keep the ordering stable, so we hash both entries
+			// and sort based on that (since end times could be equal too, in principle).
+			ah := xxhash.Sum64String(a.String())
+			bh := xxhash.Sum64String(b.String())
+
+			if ah == bh {
+				// WTF.
+				t.Logf("identical timings detected!")
```
Comment (johnstcn) [inline provisioner/terraform/internal/timings_test_utils.go:95]: Should this fail the test outright?

File: provisionersdk/proto/provisioner.proto [hunk 26]
```diff
@@ -266,6 +268,7 @@ message PlanComplete {
     repeated Resource resources = 2;
     repeated RichParameter parameters = 3;
     repeated ExternalAuthProviderResource external_auth_providers = 4;
+    repeated Timing timings = 6;
```
Comment (johnstcn) [inline provisionersdk/proto/provisioner.proto:271]: If I'm not mistaken, adding new protobuf fields here will mean that folks will need to bump external provisioners at the same time as coder?

Comment (dannykopping) [reply provisionersdk/proto/provisioner.proto:271]: AFAIK all proto fields are nilable for BC, and additional fields should be ignored; but I will validate that.
Good call 👍

File: provisioner/terraform/testdata/timings-aggregation/fake-terraform.sh
Comment (johnstcn) [inline provisioner/terraform/testdata/timings-aggregation/fake-terraform.sh:1]: nice!

Potentially later on we could just extract the Terraform outputs to some golden files and just echo those?

**Review by johnstcn — APPROVED** [review]
I don't have any blocking changes here, but it would be good to verify the situation with external provisioners on upgrade.

File: provisionersdk/proto/provisioner.proto (hunk 26 above)
Comment (dannykopping) [reply provisionersdk/proto/provisioner.proto:271]: Validated 👍

File: provisioner/terraform/testdata/timings-aggregation/fake-terraform.sh
Comment (dannykopping) [reply provisioner/terraform/testdata/timings-aggregation/fake-terraform.sh:1]: What would the value of that be (not asking rhetorically)? We probably wouldn't be able to generate the outputs, so I don't know why they would be considered golden in this context.

Comment (johnstcn) [reply provisioner/terraform/testdata/timings-aggregation/fake-terraform.sh:1]: It might make it easier to fake different outputs, but we probably don't need this right now.

File: provisionersdk/proto/provisioner.proto (hunk 26 above)
Comment (dannykopping) [reply provisionersdk/proto/provisioner.proto:271]: For future reference, here's how I validated:

- added a log line in the provisioner code to show that timing data was captured
- `make build-fat` from this branch, to produce bins capable of running `provisionerd start --psk=hellocoderd`
- switched to `main`
- ran `develop.sh` with `CODER_PROVISIONER_DAEMON_PSK=hellocoderd` and `CODER_PROVISIONER_DAEMONS=0` to force all builds to the external provisioner
- added our internal enterprise license to allow external provisioners
- created a workspace
- all worked out fine
- switched back to this branch
- ran `develop.sh` with the same args
- created another workspace
- timings were set as expected

File: coderd/provisionerdserver/provisionerdserver.go (hunk 24 above)
Comment (dannykopping) [reply coderd/provisionerdserver/provisionerdserver.go:1451]: Good call. However, I'm wondering how valuable this log entry will be without the `resource`, which is the most likely attribute which may leak data. I think Terraform is pretty good with keeping sensitive strings masked, so I'll defer to their good judgment for now. If we were running a SaaS, I'd be more paranoid.

File: provisionerd/proto/provisionerd.pb.go (hunk 11 above)
Comment (dannykopping) [reply provisionerd/proto/provisionerd.pb.go]: Added https://github.com/coder/coder/issues/14343 as a follow-up.
